### PR TITLE
case rebuild for sql

### PR DESCRIPTION
**Pull request by snopoke**
@benrudolph @dannyroberts @czue @NoahCarnahan 
This one could use a proper review along with https://github.com/dimagi/commcare-hq/pull/9100

:blowfish:

### CODE REVIEW

File: corehq/form_processor/backends/sql/processor.py [hunk 1]
```diff
@@ -169,3 +173,87 @@ def xformerror_from_xform_instance(cls, instance, error_message, with_new_id=Fal
...
+    def get_form(form_id):
+        if form_id in updated_xforms_map:
+            return updated_xforms_map[form_id]
+
+        try:
+            return xform_map[form_id]
+        except KeyError:
+            raise XFormNotFound
+
+    for transaction in transactions:
```
Comment (dimagimon) [inline corehq/form_processor/backends/sql/processor.py]: F402 import 'transaction' from line 6 shadowed by loop variable

File: corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py [hunk 2]
```diff
@@ -167,7 +170,7 @@ def _confirm_action_order(case, expected_actions):
         _confirm_action_order(case, [a1, a2, a3])
 
     def testRebuildEmpty(self):
-        self.assertEqual(None, rebuild_case_from_forms('notarealid'))
+        self.assertEqual(None, rebuild_case_from_forms('anydomain', 'notarealid', RebuildWithReason(reason='test')))
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py]: E501 line too long (116 > 115 characters)

File: corehq/form_processor/backends/sql/processor.py [hunk 3]
```diff
@@ -169,3 +176,93 @@ def xformerror_from_xform_instance(cls, instance, error_message, with_new_id=Fal
...
+                    deprecated_form = xform
+                affected_cases.update(case_update.id for case_update in get_case_updates(xform))
+
+            rebuild_detail = FormEditRebuild(deprecated_form_id=deprecated_form.form_id)
+            for case_id in affected_cases:
+                case = case_db.get(case_id)
+                if not case:
+                    case = CommCareCaseSQL(domain=domain, case_id=case_id)
+                    case_db.set(case_id, case)
+                case = FormProcessorSQL._rebuild_case_from_transactions(case, rebuild_detail, updated_xforms=xforms)
```
Comment (dimagimon) [inline corehq/form_processor/backends/sql/processor.py]: E501 line too long (116 > 115 characters)

File: corehq/form_processor/backends/sql/update_strategy.py [hunk 4]
```diff
@@ -1,8 +1,13 @@
 import logging
+
+from datetime import datetime
```
Comment (dimagimon) [inline corehq/form_processor/backends/sql/update_strategy.py]: F401 'datetime' imported but unused

**Comment by snopoke** [comment]
:white_check_mark: finally

**Comment by benrudolph** [comment]
holy guacamole! hell of PR. happy to merge, do you want to wait for others to review?

**Comment by czue** [comment]
(no need to block on me - my plan was to do it on the design review)

**Comment by benrudolph** [comment]
i think i'll go ahead and merge then and assume that everyone else will review during the design review. no need to let this bit rot

**Pull request merged**